apm-sniffer/apm-agent-core/src/main/java/org/apache/skywalking/apm/agent/core/conf/SnifferConfigInitializer.java

```diff
@@ -234,4 +235,12 @@ static void configureLogger() {
                 LogManager.setLogResolver(new PatternLogResolver());
         }
     }
+
+    public static void pluginInitCompleted() {
+        IS_PLUGIN_INIT_COMPLETED = true;
+    }
+
+    public static boolean isPluginInitCompleted() {
+        return IS_PLUGIN_INIT_COMPLETED;
+    }
```

Comment:
What are these about?

Reply:
the `FileWriter` class in Skywalking agent uses `ThreadPoolExecutor`. If plugins are not loaded, FileWriter is created. Then `ThreadPoolExecutor won't be instructed`. so we must make sure that after plugins are loaded, FileWriter can be created.

```
    private FileWriter() {
        logBuffer = new ArrayBlockingQueue(1024);
        final ArrayList<String> outputLogs = new ArrayList<String>(200);
        Executors.newSingleThreadScheduledExecutor(new DefaultNamedThreadFactory("LogFileWriter"))
                 .scheduleAtFixedRate(new RunnableWithExceptionProtection(new Runnable() {
                     @Override
                     public void run() {
                         try {
                             logBuffer.drainTo(outputLogs);
                             for (String log : outputLogs) {
                                 writeToFile(log + Constants.LINE_SEPARATOR);
                             }
                             try {
                                 fileOutputStream.flush();
                             } catch (IOException e) {
                                 e.printStackTrace();
                             }
                         } finally {
                             outputLogs.clear();
                         }
                     }
                 }, new RunnableWithExceptionProtection.CallbackWhenException() {
                     @Override
                     public void handle(Throwable t) {
                     }
                 }), 0, 1, TimeUnit.SECONDS);
    }
```

Reply:
But simply change like this, you would make a lot of logs disappear from `agent.log`

Reply:
Yeah, Before plugins are loaded completed, logs will be printed by system.out.

Reply:
Also, if you need a new flag of plugin initialization, it looks like the flag should be in `PluginFinder`? [SnifferConfigInitializer.java](https://github.com/apache/skywalking-java/pull/146/files/22e85142ff56493cee90adf49d123b71ff3e0089#diff-3f4c9264a89197350b7eb2ca0e2f2daa494306980a00142edef7f5abfc941951) seems a wrong place to add.

Reply:
@Cool-Coding I think this part of codes has not been polished yet.

Reply:
> @Cool-Coding I think this part of codes has not been polished yet.

I have polished the `WriterFactoryTest` UT adding the code `        BDDMockito.given(SnifferConfigInitializer.isPluginInitCompleted()).willReturn(true)`. The UT passed.

And the statement `SnifferConfigInitializer.pluginInitCompleted()` has been moved after `installOn(instrumentation)`.

```
        agentBuilder.type(pluginFinder.buildMatch())
                    .transform(new Transformer(pluginFinder))
                    .with(AgentBuilder.RedefinitionStrategy.RETRANSFORMATION)
                    .with(new RedefinitionListener())
                    .with(new Listener())
                    .installOn(instrumentation);

        SnifferConfigInitializer.pluginInitCompleted();

        try {
            ServiceManager.INSTANCE.boot();
        } catch (Exception e) {
            LOGGER.error(e, "Skywalking agent boot failure.");
        }
```

Reply:
> Also, if you need a new flag of plugin initialization, it looks like the flag should be in PluginFinder? SnifferConfigInitializer.java seems a wrong place to add.

I mean which class should own the flag. `SnifferConfigInitializer` has nothing related to the plugin mechanism.

Reply:
Same. I am not arguing about the process flow, I am talking about where to hold this flag.

I mean moving this flag to https://github.com/apache/skywalking-java/pull/146#discussion_r848061270